api/reporting/schemas/REPORTING-SCHEMA-TOOLS.md

```diff
@@ -0,0 +1,177 @@
+and tools to generate and translate them into code. There
+are a few different stages, but the grunt work is mostly automated
+for repeatability and consistency.
+
+The ruby script `generate_all_tables.rb` will operate on all of the input CSV files (described below)
+and produce output files in this directory structure 
+```
+$ ls ~/scratch/reportingOut5
+big_query_json       projection_interface projection_query     swagger_yaml
+```
+
+## EXPLAIN Application DB Table
+The `explain table` MySql [statement](https://dev.mysql.com/doc/refman/8.0/en/explain.html) generates a tabular
+set of attributes for table. These are then saved as CSV files and placed into the `mysql_describe_csv` directory.
+
+For example, the output of the `explain address;` is shown below.
+
+| Field | Type | Null | Key | Default | Extra |
+| :--- | :--- | :--- | :--- | :--- | :--- |
+| id | bigint\(20\) | NO | PRI | NULL | auto\_increment |
+| street\_address\_1 | varchar\(95\) | NO |  | NULL |  |
+| street\_address\_2 | varchar\(95\) | YES |  | NULL |  |
+| zip\_code | varchar\(10\) | YES |  | NULL |  |
+| city | varchar\(95\) | NO |  | NULL |  |
+| state | varchar\(95\) | NO |  | NULL |  |
+| country | varchar\(95\) | NO |  | NULL |  |
+| user\_id | bigint\(20\) | NO | MUL | NULL |  |
+
+## Mapping to BigQuery Schema JSON Format
+The next task is to translate the MySql table description into the [BigQuery schema format](https://cloud.google.com/bigquery/docs/schemas#specifying_a_json_schema_file). Given this CSV input, we only really care about the field name and types. The
+relational constraints won't be preserved in BigQuery, so they're basically ignored,
+but we do want the primary key (which would be renamed to `address_id` in this example,
+though in this case we actually merge `address` columns into the
+`user` table). The script generates the for each table, with the example
+of the address table shown. All fields are marked nullable, as to do otherwise would involve
+adding constraints on the source data and/or its snapshot & upload processes.
+
+```json
+[
+  {
+    "name": "id",
+    "description": "",
+    "type": "INT64",
+    "mode": "NULLABLE"
+  },
+  {
+    "name": "street_address_1",
+    "description": "",
+    "type": "STRING",
+    "mode": "NULLABLE"
+  },
+  {
+    "name": "street_address_2",
+    "description": "",
+    "type": "STRING",
+    "mode": "NULLABLE"
+  },
+  {
+    "name": "zip_code",
+    "description": "",
+    "type": "STRING",
+    "mode": "NULLABLE"
+  },
+  {
+    "name": "city",
+    "description": "",
+    "type": "STRING",
+    "mode": "NULLABLE"
+  },
+  {
+    "name": "state",
+    "description": "",
+    "type": "STRING",
+    "mode": "NULLABLE"
+  },
+  {
+    "name": "country",
+    "description": "",
+    "type": "STRING",
+    "mode": "NULLABLE"
+  },
+  {
+    "name": "user_id",
+    "description": "",
+    "type": "INT64",
+    "mode": "NULLABLE"
+  }
+]
+```
+
+## Spring Data Projection Interface
```

Comment:
A prototype example of this is in [this PR](https://github.com/all-of-us/workbench/pull/3949/files#diff-8d7ae350997a1bb426a90e2e30b11020R73) (changing rapidly).